### PR TITLE
Improvements to unbounded enum/bool ranges

### PR DESCRIPTION
**Pull request by bradcray**
This PR continues the work of #19991 by making partially bounded or
unbounded ranges of enum/bool work more similarly to fully bounded
ranges using the extreme values of the bool/enum in cases like
iteration.  In other cases like slicing, they still behave differently.  For
example:

```chapel
enum color {red, green, blue};
use color;
for c in red..  // this will iterate over all three colors then stop, similar to `red..blue`
var A: [red..green] real;
A[red..]  // this is equivalent to A[red..green], as with other unbounded slicing cases, inheriting A's upper bound
```

It also fixes some cases pointed out by Lydia in the previous PR such as
`for c in ..blue` not working properly, and implements range equality
such that `red.. == red..blue` now returns true as Vass suggested.

The main changes here involved updating lots of logic in the range module
from making decisions based on whether the static type of the range is fully
bounded to whether or not the range has well-defined upper and lower bounds.
For example:

```chapel
red..blue  // statically fully bounded, so has upper and lower bounds
red..     // statically unbounded on the upper bound, but still has upper and lower bounds for the sake of things like iteration
```

This also required some code refactoring to avoid querying low/high bounds
of ranges that didn't have them defined well.  A good follow-on step to make
sure this is airtight would be to optimize away the unnecessary `_low` and
`_high` fields from unbounded ranges to ensure that we're not referring to
them incorrectly (I found some cases where we were, but others likely remain).
In some cases, additional care also had to be taken to keep default ranges
over single-value enums working as before.

Other minor changes:
* refined wording about iterating over unbounded ranges since `red..`
  works but is technically unbounded
* fixed default unbounded ranges over bool
* fixed striding single-enum ranges by 1 (?)

TODO:
- [ ] get design approval
- [ ] file follow-up issue about optimizing away _low/_high fields for safety
- [ ] file follow-up design issue about isBoundedRange() which seems increasingly... vague
- [ ] consider whether futures that were failing after #19991 was retired can now be simplified or need further care

### CODE REVIEW

**Comment by bradcray** [comment]
@vasslitvinov : I think this is now ready to review if you have time to give it a look.  @lydia-duncan may also be interested since it addresses an issue she identified in the demo of #19991.

File: test/types/range/unbounded/compareEnumBool.chpl [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+use color;
+
+var rgb = red..blue;
+var rdd = red..;
+var ddb = ..blue;
+var empty: range(color, BoundedRangeType.boundedNone) = ..;
+
+writeln(rgb == rdd);
+writeln(rgb == ddb);
+writeln(rgb == empty);
```
Comment (lydia-duncan) [inline test/types/range/unbounded/compareEnumBool.chpl:11]: I don't know that empty is the best name for this, but I wouldn't hold up the PR over it

**Review by lydia-duncan — APPROVED** [review]
I took a look over just the test cases and they seem reasonable to me.  Just called out the naming of one variable but it's probably fine to leave it as is

File: modules/internal/ChapelRange.chpl [hunk 2]
```diff
@@ -573,7 +613,17 @@ module ChapelRange {
...
+      }
+    }
+    if (boundedType == BoundedRangeType.boundedLow ||
+        boundedType == BoundedRangeType.boundedNone) {
+      if isEnumType(idxType) {
+        return chpl_intToIdx(idxType.size - 1);
+      } else if isBoolType(idxType) {
+        return chpl_intToIdx(1);
+      } else {
+        compilerError("Querying the high bound of a range not thought to have one");
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl:626]: Please consider making this message more formal, less colloquial.

File: modules/internal/ChapelRange.chpl [hunk 3]
```diff
@@ -762,7 +821,7 @@ module ChapelRange {
      partially bounded, the return value will not (cannot) be a
      `param`.
   */
-  proc range.hasFirst() param where !stridable && !hasHighBound()
+  proc range.hasFirst() param where !stridable && !(hasLowBound() && hasHighBound())
     return hasLowBound();
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl:825]: I think this overload's where clause should be simply `where !stridable`.

File: modules/internal/ChapelRange.chpl [hunk 4]
```diff
@@ -792,7 +851,7 @@ module ChapelRange {
      partially bounded, the return value will not (cannot) be a
      `param`.
   */
-  proc range.hasLast() param where !stridable && !hasLowBound()
+  proc range.hasLast() param where !stridable && !(hasLowBound() && hasHighBound())
     return hasHighBound();
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl:855]: I think this overload's where clause should be simply `where !stridable`.

File: modules/internal/ChapelRange.chpl [hunk 5]
```diff
@@ -830,17 +889,21 @@ module ChapelRange {
      if the represented sequence is not defined. */
   inline proc range.contains(ind: idxType)
   {
+    if chpl__singleValIdxType(idxType) {
+      if _low > _high then return false;
+    }
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl:894]: Makes me want to have `proc range.isEmptySingleValueRange()`, as well as possibly (for other uses) `range.isSingleValueRange()`

File: modules/internal/ChapelRange.chpl [hunk 6]
```diff
@@ -2601,10 +2684,12 @@ operator :(r: range(?), type t: range(?)) {
...
     var ret: string;
 
-    if x.hasLowBound() then
+    if (x.boundedType == BoundedRangeType.bounded ||
+        x.boundedType == BoundedRangeType.boundedLow) then
       ret += x.lowBound:string;
     ret += "..";
-    if x.hasHighBound() {
+    if (x.boundedType == BoundedRangeType.bounded ||
+        x.boundedType == BoundedRangeType.boundedHigh) {
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl:2692]: Perhaps `x.hasDeclaredHighBound()` and ditto low bound?

**Review by vasslitvinov — APPROVED** [review]
This PR adds code in `range.init=` that seems necessary because we can initialize an enum range from an int range. My preference would be to disallow such initialization, given that we cannot initialize an enum variable from an integer.

While this does not affect this PR, please mark this as a follow-up item, unless in the past we made a concious choice to allow range(color).init=(range(int)).

In the PR message:

> In other cases like slicing, they still behave differently.

I do not see this. `A[red..]` does not illustrate this. Please provide an example or rephrase the statement.

> file follow-up design issue about isBoundedRange() 

Please discuss also hasLowBound() and hasHighBound() on that issue.

Overall looks good, see my comments.

File: modules/internal/ChapelRange.chpl [hunk 7]
```diff
@@ -152,6 +160,13 @@ module ChapelRange {
   proc range.init() {
     this.idxType = int;
     this.boundedType = BoundedRangeType.boundedNone;
+    this.complete();
+    if isBoolType(idxType) || isEnumType(idxType) then
+      this._low = 0;
+    if isBoolType(idxType) then
+      this._high = 1;
+    else if isEnumType(idxType) then
+      this._high = idxType.size-1;
```
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:165]: Refactored this in 109fe8ab780d0a7913290a9ca3868a308db08489.

File: modules/internal/ChapelRange.chpl (hunk 2 above)
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:640]: Done in 0a245b81df611da5202e59316d7cf61e36466fb0

File: modules/internal/ChapelRange.chpl (hunk 6 above)
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:2803]: Nice idea.  Addressed this in 646ef84876a1096151f322932f29537bd9766ede.

File: modules/internal/ChapelRange.chpl (hunk 4 above)
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:875]: I'd hoped this during development as well, but don't believe it's the case (at least not without other additional changes).  For example, making this change causes `(4..3).hasFirst()` or `.hasLast()` to return `true` where the correct answer is `false`.

File: modules/internal/ChapelRange.chpl (hunk 3 above)
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:846]: See conversation below on range.hasLast().

**Comment by bradcray** [comment]
As a note for myself and anyone following this:  As I've been thinking about running the design proposal for the changes here and we've been focused on ranges in MS19, I've been finding myself of two minds about whether `false..` or `red..` should support a `.high` or `.highBound` query given that their type is essentially "I have no high bound" (and we want them to continue to behave that way in the slicing context).  That said, having them have a `first` and `last` seems much more reasonable (in the sense of "If you were to iterate over me, what would you get?").  I'll likely end up opening up an issue about this even though my preference would just be to bite the bullet and proceed given how long it's taken me to land this PR (due to lack of time).

Also, while writing more tests, I found that even though `for c in (red..)` works fine, `for c in zip(red.., 1..)` does not due to #20051, so I've started a PR to try and address that independently of this PR.

**Comment by bradcray** [comment]
I've opened #20896 as a design issue to make sure what I've done here seems right, where I'm a bit skeptical that it is when it comes to high/low bounds queries

**Comment by bradcray** [comment]
Closing in favor of #21332

**Pull request closed without merging**